components/homme/src/share/reduction_mod.F90

```diff
@@ -337,6 +342,92 @@ subroutine pmax_mt_r_1d(red,redp,len,hybrid)
   end subroutine pmax_mt_r_1d
 
 
+!max with location index
+!at the end result is in redp thread local array
+  subroutine ParallelMaxWithIndex(redp,hybrid)
+    use hybrid_mod, only : hybrid_t
+#ifdef _MPI
+    use parallel_mod, only: mpi_maxloc, mpi2real_t
+#endif
+
+    real (kind=real_kind), intent(inout) :: redp(2) ! thread private vector of partial sum
+    type (hybrid_t),       intent(in)    :: hybrid  ! parallel handle
+
+    ! Local variables
+    integer ierr, k
+
+    ! the first and fastest thread performs initializing copy
+    !$OMP SINGLE
+    red_max_index%buf(1:2) = redp(1:2)
+    red_max_index%ctr = hybrid%ithr
+    !$OMP END SINGLE
+
+    ! all other threads now do the max_op wrt the first thread's data
```

Comment:
Perhaps: "all threads now" (no "other") since the thread that ran the SINGLE code will still do this critical section, too.